File: AtCoder/ABC/ABC406/F.py
import sys
input = sys.stdin.readline

N = int(input())
Ad = [[] for _ in range(N + 1)]
U = [0] * N
V = [0] * N
for i in range(1, N):
  u, v = map(int, input().split())
  U[i], V[i] = u, v
  Ad[u].append(v)
  Ad[v].append(u)

t = 0
E = [0] * (N + 1)
X = [0] * (N + 1)
P = [0] * (N + 1)
que = [(1, 0, 0)]
while que:
  cx = que.pop()
  u, p, stat = cx
  if stat == 0:
    P[u] = p
    t += 1
    E[u] = t
    que.append((u, p, 1))
    for w in Ad[u]:
      if w != p:
        que.append((w, u, 0))
  else:
    X[u] = t

for i in range(1, N):
  u, v = U[i], V[i]
  if P[u] == v:
    U[i] = u
  else:
    U[i] = v

bits = [0] * (N + 1)
for i in range(1, N + 1):
  bits[i] = i & -i
total = N

q = int(input())
out = []
for _ in range(q):
  line = input().split()
  if line[0] == '1':
    x = int(line[1])
    w = int(line[2])
    total += w
    idx = E[x]
    while idx <= N:
      bits[idx] += w
      idx += idx & -idx
  else:
    y = int(line[1])
    c = U[y]
    l = E[c]
    r = X[c]
    s, t = 0, 0
    i2 = r
    while i2 > 0:
      s += bits[i2]
      i2 -= i2 & -i2
    i2 = l - 1
    while i2 > 0:
      t += bits[i2]
      i2 -= i2 & -i2
    rr = s - t
    diff = total - 2 * rr
    if diff < 0:
      diff = -diff
    out.append(str(diff))

print('\n'.join(out))

File: AtCoder/ABC/ABC410/F.cpp
// PythonではO(HW^2)が間に合わない
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using VV = vector<vector<int>>;

void solve() {
    int H, W;
    cin >> H >> W;
    VV A(H, vector<int>(W));

    for (int i = 0; i < H; ++i) {
        string s;
        cin >> s;
        for (int j = 0; j < W; ++j) {
            A[i][j] = (s[j] == '#') ? 1 : -1;
        }
    }

    // 行数が列数を超えるときは転置
    if (H > W) {
        VV B(W, vector<int>(H));
        for (int i = 0; i < H; ++i)
            for (int j = 0; j < W; ++j)
                B[j][i] = A[i][j];
        A.swap(B);
        swap(H, W);
    }

    ll ans = 0;

    vector<int> freq(2 * H * W + 1);
    vector<int> used;
    used.reserve(W + 1);

    // 上端u, 下端dを全探索
    for (int u = 0; u < H; ++u) {
        vector<int> C(W, 0);
        for (int d = u; d < H; ++d) {
            // 各列の累積和を更新
            for (int j = 0; j < W; ++j) {
                C[j] += A[d][j];
            }

            // freq と used をリセット（計算量を抑えるため）
            used.clear();
            freq[H * W] = 1;
            used.push_back(H * W);

            ll prefix = 0;
            for (int j = 0; j < W; ++j) {
                prefix += C[j];
                int idx = int(prefix) + H * W;
                ans += freq[idx];
                if (freq[idx] == 0) {
                    used.push_back(idx);
                }
                freq[idx]++;
            }

            for (int idx : used) {
                freq[idx] = 0;
            }
        }
    }
    cout << ans << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        solve();
    }
    return 0;
}

/*
def solve():
    import sys
    from collections import defaultdict
    input = sys.stdin.readline

    H, W = map(int, input().split())
    A = [[1 if c == '#' else -1 for c in input().rstrip()] for _ in range(H)]

    if H > W:
        A = list(map(list, zip(*A)))
        H, W = W, H

    ans = 0

    for u in range(H):
        C = [0] * W

        for d in range(u, H):
            for j in range(W):
                C[j] += A[d][j]

            pp = 0
            fq = defaultdict(int)
            fq[0] = 1

            for j in range(W):
                pp += C[j]
                ans += fq[pp]
                fq[pp] += 1
    
    print(ans)

def main():
    T = int(input())        
    for _ in range(T):
        solve()

if __name__ == "__main__":
    main()

*/